pydantic_ai_slim/pydantic_ai/_function_schema.py

```diff
@@ -127,6 +126,9 @@ def function_schema(  # noqa: C901
             errors.append(f'Missing parameter descriptions for {", ".join(missing_params)}')
 
     for index, (name, p) in enumerate(sig.parameters.items()):
+        if index == 0 and takes_ctx is None:
+            takes_ctx = p.annotation is not sig.empty and _is_call_ctx(type_hints[name])
```

Comment:
Instead of using `_takes_ctx()`, do the logic here. This avoids computing the type hints twice.